File: Lab4.1-Rational-Numbers/Lab4/Rational.cpp
//
//  Rational.cpp
//  Lab4
//
//  Created by Moeen Khan on 11/6/20.
//
//
//#include "Rational.hpp"
//#ifndef RATIONAL_H
//#define RATIONAL_H


#include "Rational.hpp"
#include<iostream>
#include<cstdlib>
#include<string>
#include "RationalException.h"
#include <iostream>
#include <sstream>
#include <string>
#include <ctime>
#include "gcd.h"
using namespace std;

Rational::Rational(int num, int denom)
{
    if(denom == 0) throw RationalException("Denominator cannot be 0");

    normalize(num, denom);

    Rational::num = num;
    Rational::denom = denom;
}


int Rational::getNumerator() const { return num; }
int Rational::getDenominator() const { return denom; }


Rational &Rational::addInPlace(const Rational &r)
{
    int commonD = denom * r.getDenominator();
    int numerator1 = num * r.getDenominator();
    int numerator2 = r.getNumerator() * denom;

    num = numerator1 + numerator2;
    denom = commonD;
    normalize(num, denom);
    
    return *this;
}

//Rational &Rational::subInPlace(const Rational &r) { return addInPlace(r.neg()); }

Rational &Rational::subInPlace(const Rational &r){
    addInPlace(r.neg());
    return *this;
}

Rational &Rational::mulInPlace(const Rational &r)
{
    num *= r.num;
    denom *= r.getDenominator();
    normalize(num, denom);

    return *this;
}

//Rational &Rational::divInPlace(const Rational &r) { return mulInPlace(r.inv()); }

Rational &Rational::divInPlace(const Rational &r){
    mulInPlace(r.inv());

    return *this;
}

//Rational Rational::add(const Rational &r) const
//{
//    Rational newR(num, denom);
//    return newR.addInPlace(r);
//}

Rational Rational::add(const Rational &r) const{
    Rational rational(num, denom);
    rational.addInPlace(r);
    return rational;
}

//Rational Rational::sub(const Rational &r) const
//{
//    Rational newR(num, denom);
//    return newR.subInPlace(r);
//}

Rational Rational::sub(const Rational &r) const{
    Rational rational(num, denom);
    rational.subInPlace(r);
    return rational;
}

//Rational Rational::mul(const Rational &r) const
//{
//    Rational newR(num, denom);
//    return newR.mulInPlace(r);
//}

Rational Rational::mul(const Rational &r) const{
    Rational rational(num, denom);
    rational.mulInPlace(r);

    return rational;
}

//Rational Rational::div(const Rational &r) const
//{
//    Rational newR(num, denom);
//    return newR.divInPlace(r);
//}

Rational Rational::div(const Rational &r) const{
    Rational rational(num, denom);
    rational.divInPlace(r);
    return rational;
}

 

void normalize(int &num, int &denom){
    int g = gcd(num, denom);
    
    num /= g;
    denom /= g;
}



Rational Rational::neg() const {
    return Rational(num * -1, denom);
}
Rational Rational::inv() const {
    return Rational(denom, num);
}

bool Rational::equals(const Rational &r) const {
    return this->compareTo(r) == 0;
}

int Rational::compareTo(const Rational &r) const
{
    int lhs = num * r.getDenominator();
    int rhs = denom * r.getNumerator();
    return lhs < rhs ? -1 : (lhs > rhs ? 1 : 0);
}

File: Lab2.3 - Rational Number/src/Rational.java
public class Rational
{
    private int numerator;
    private int denominator;

    public Rational(int numerator, int denominator) throws RationalException {
        if (denominator == 0)
            throw new RationalException("Denominator can not be Zero");

        this.numerator = numerator;
        this.denominator = denominator;

        normalize();

    }

    public Rational(int numerator) {
        this(numerator, 1);
    }

    public Rational() {
        this(0,1);
    }

    Rational(Rational rational) {
        this.numerator = rational.numerator;
        this.denominator = rational.denominator;
    }


    public Rational negate(){
        return new Rational(-1*this.numerator, this.denominator);
    }


    public Rational inverse(){
        return new Rational(this.denominator, this.numerator);
    }

    public void addInPlace(Rational r)
    {
        int commonDenom = this.denominator * r.getDenominator();
        int newNum1 = this.numerator * r.getDenominator();
        int newNum2 = r.getNumerator() * this.denominator;

        this.numerator = newNum1 + newNum2;
        this.denominator = commonDenom;
        normalize();
    }

    public Rational add(Rational r)
    {
        Rational newR = new Rational(this);
        newR.addInPlace(r);
        return newR;
    }

    public void subInPlace(Rational r)
    {
        this.addInPlace(r.negate());
    }

    public Rational sub(Rational r)
    {
        Rational newR = new Rational(this);
        newR.subInPlace(r);
        return newR;
    }

    public void mulInPlace(Rational r)
    {
        this.numerator *= r.getNumerator();
        this.denominator *= r.getDenominator();
        normalize();
    }

    public Rational mul(Rational r)
    {
        Rational newR = new Rational(this);
        newR.mulInPlace(r);
        return newR;
    }


    public void divInPlace(Rational r)
    {
        this.mulInPlace(r.inverse());
    }

    public Rational div(Rational r)
    {
        Rational newR = new Rational(this);
        newR.divInPlace(r);
        return newR;
    }

    public int getNumerator() { return this.numerator; }

    public int getDenominator() { return this.denominator; }


    public boolean equals(Rational rational){

        if(this.numerator == rational.numerator && this.denominator == rational.denominator)
            return true;
        else
            return false;
    }


    public int compareTo(Rational rational){

        double thisVal = (double) numerator / denominator;
        double rationalVal = (double) rational.numerator/ rational.denominator;

        if(rationalVal < thisVal)
            return 1;

        else if(rationalVal == thisVal)
            return 0;

        else
            return -1;
    }

    private int gcd(int numerator, int denominator){
        if (denominator == 0)
            return numerator;

        else
            return gcd(denominator, numerator%denominator);
    }

    private void normalize()
    {
        int gcd = gcd(this.numerator, this.denominator);
        this.numerator /= gcd;
        this.denominator /= gcd;
    }

        @Override
    public String toString() {
        if(denominator == 0 || denominator == 1)
            return Integer.toString(numerator);

        else
        return numerator + "/" + denominator;

    }

}


